hw/ip/prim/dv/prim_ram_scr/cpp/scramble_model.cc

```diff
@@ -0,0 +1,373 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "scramble_model.h"
+#include <cassert>
+#include <iostream>
+#include <stdint.h>
+#include <vector>
+#include "prince_ref.h"
+
+uint8_t PRESENT_SBOX4[] = {0xc, 0x5, 0x6, 0xb, 0x9, 0x0, 0xa, 0xd,
+                           0x3, 0xe, 0xf, 0x8, 0x4, 0x7, 0x1, 0x2};
+
+uint8_t PRESENT_SBOX4_INV[] = {0x5, 0xe, 0xf, 0x8, 0xc, 0x1, 0x2, 0xd,
+                               0xb, 0x4, 0x6, 0x3, 0x0, 0x7, 0x9, 0xa};
+
+static const uint32_t kNumAddrSubstPermRounds = 2;
+static const uint32_t kNumDataSubstPermRounds = 2;
+static const uint32_t kNumPrinceHalfRounds = 2;
+
+static std::vector<uint8_t> byte_swap_vector(
+    const std::vector<uint8_t> &vec_in) {
+  std::vector<uint8_t> vec_out(vec_in.size());
+
+  for (int i = 0; i < vec_in.size(); ++i) {
+    vec_out[vec_in.size() - i - 1] = vec_in[i];
+  }
+
+  return vec_out;
+}
+
+static uint8_t read_vector_bit(const std::vector<uint8_t> &vec,
+                               uint32_t bit_pos) {
+  assert(bit_pos / 8 < vec.size());
+
+  return (vec[bit_pos / 8] >> (bit_pos % 8)) & 1;
+}
+
+static void or_vector_bit(std::vector<uint8_t> &vec, uint32_t bit_pos,
+                          uint8_t bit) {
+  assert(bit_pos / 8 < vec.size());
+
+  vec[bit_pos / 8] |= bit << (bit_pos % 8);
+}
+
+// Run each 4-bit chunk of bytes from `in` through the SBOX. Where `bit_width`
+// isn't a multiple of 4 the remaining bits are just copied straight through.
+// `invert` choose whether to use the inverted SBOX or not.
+static std::vector<uint8_t> scramble_sbox_layer(const std::vector<uint8_t> &in,
+                                                uint32_t bit_width,
+                                                bool invert) {
+  assert(in.size() >= ((bit_width + 7) / 8));
```

Comment:
I think it might be better to make this assertion stronger, claiming that `in` is the right number of bytes to hold `bit_width` bits. Any reason not to replace the `>=` with `==`? Same for the other occurrences of this assertion below.

Reply:
It's to allow use of vectors that have more bytes in than you're actually using.

Reply:
Do we have any code that does this? If not, the assertion should be stronger, I think. After all, it would be kind of weird to do things like bit-reversing just the first part of a vector, probably indicating a programming error.

Reply:
We do, there's some code in `scrambled_ecc32_mem_area.cc` that always produces 4 byte vectors from an integer address. However it was quite easy to adjust so I've hardened up all the assertions as you suggest.